openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,333 @@
+
+    # First validate each preference setting
+    errors = {}
+    for preference_key in update.keys():
+        preference_value = update[preference_key]
+        if preference_value is not None:
+            try:
+                _validate_user_preference(existing_user, preference_key, preference_key)
+            except PreferenceValidationError as error:
+                developer_message = error.preference_errors[preference_key]["developer_message"]
+                errors[preference_key] = {
+                    "developer_message": developer_message
+                }
+    if errors:
+        raise PreferenceValidationError(errors)
+    # Then perform the patch
+    try:
+        for preference_key in update.keys():
+            preference_value = update[preference_key]
+            if preference_value is not None:
+                UserPreference.set_preference(existing_user, preference_key, preference_value)
+            else:
+                UserPreference.delete_preference(existing_user, preference_key)
+    except Exception as error:
+        raise PreferenceUpdateError(error.message)
+
+
+@intercept_errors(UserAPIInternalError, ignore_errors=[UserAPIRequestError])
+@transaction.commit_on_success
+def set_user_preference(requesting_user, preference_key, preference_value, username=None):
+    """Update a user preference for the given username.
+
+    Note:
+        It is up to the caller of this method to enforce the contract that this method is only called
+        with the user who made the request.
+
+    Arguments:
+        requesting_user (User): The user requesting to modify account information. Only the user with username
+            'username' has permissions to modify account information.
+        preference_key (string): The key for the user preference.
+        preference_value (string): The value to be stored. Non-string values will be converted to strings.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        UserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        UserNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        PreferenceValidationError: the update was not attempted because validation errors were found
+        PreferenceUpdateError: the operation failed when performing the update.
+        UserAPIInternalError: the operation failed due to an unexpected error.
+    """
+    existing_user = _get_user(requesting_user, username)
+    if preference_value is None or preference_value == '':
+        message = _('Preference {preference_key} cannot be set to an empty value').format(
+            preference_key=preference_key
+        )
+        raise PreferenceUpdateError(message, user_message=message)
```

Comment:
This should be a ValidationError instead (and make sure there is test coverage for it).